### PR TITLE
Add validate hook for index-type validate.

### PR DESCRIPTION
**Pull request by MikeStall**
Validator hook is dispatched after the %% are resolved, but before any runtime invocation (so no { } resolved yet).

### CODE REVIEW

**Comment by mathewc** [comment]
Can you provide more context? What is the scenario here? I don't understand the above description.

**Comment by MikeStall** [comment]
This is a hook to validate static things in the attribute and get early errors at index time (not waiting until runtime). For example, it may validate that the properties  conforms to some regex or are in some range.
________________________________

File: src/Microsoft.Azure.WebJobs.Host/Config/ExtensionConfigContextExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    {
+        /// <summary>
+        /// Register a set of rules to handle binding to a given attribute. 
+        /// </summary>
+        /// <typeparam name="TAttribute">The attribute to bind to</typeparam>
+        /// <param name="context"></param>
+        /// <param name="validator">a validator function to invoke on the attribute before runtime. </param>
+        /// <param name="bindingProviders">list of binding providers to handle this attribute. 
+        /// If none of these handle the attribute, an error is thrown at indexing. </param>
+        public static void RegisterBindingRules<TAttribute>(
```
Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Config/ExtensionConfigContextExtensions.cs]: Related to my comments in the outstanding PR https://github.com/Azure/azure-webjobs-sdk/pull/711 I think we shouldn't use this "Rules" terminology, and instead call these what they are. E.g. RegisterBindingProviders.

Comment (mathewc) [inline src/Microsoft.Azure.WebJobs.Host/Config/ExtensionConfigContextExtensions.cs]: Also, I don't see any tests or usages of these new extension methods in this review, so it's hard to see usage context. I suggest either adding tests, or moving this into your other Queue PR and having one of the bindings use these extensions, so we can see the usage.

Comment (MikeStall) [inline src/Microsoft.Azure.WebJobs.Host/Config/ExtensionConfigContextExtensions.cs]: There are some tests in ValidationTests.cs

---
In reply to: [63563653](https://github.com/Azure/azure-webjobs-sdk/pull/723#discussion_r63563653) [](ancestors = 63563653)

**Comment by MikeStall** [comment]
This is a hook to validate static things in the attribute and get early errors at index time (not waiting until runtime). For example, it may validate that the properties  conforms to some regex or are in some range.

---
In reply to: [219741200](https://github.com/Azure/azure-webjobs-sdk/pull/723#issuecomment-219741200) [](ancestors = 219741200)

**Comment by mathewc** [comment]
closing, since I believe this is merged already?

**Pull request closed without merging**